### ISSUE TITLE
Convert all FCalls with HELPER_METHOD_FRAME to QCalls

### ISSUE THREAD

**Issue by jkotas** [dotnet/runtime#95695]
Benefits:
- Performance: QCalls are generally faster than FCalls with HELPER_METHOD_FRAME
- Eliminate duplication: These are redundant mechanism. Once all of them are converted, we can eliminate the duplication.
- Complexity of stackwalking algorithm: HELPER_METHOD_FRAMEs make the stackwalking algorithm with all its dependencies more complicated than it needs to be. (See #93550 for more context.)
   - We want to work towards unification of CoreCLR stackwalker and native AOT stackwalker. Native AOT has much simpler stackwalker and it does not have HELPER_METHOD_FRAMEs.

**Comment by hughbe** [comment]
This is cool! Is the plan to eventually remove FCall functionality from the runtime?

**Comment by davidwrighton** [comment]
I think we also might be able to move to QCalls with SuppressGCTransition instead of FCalls without HELPER_METHOD_FRAME for the perf critical ones, but that's a bridge to examine and consider crossing once we've removed HELPER_METHOD_FRAME from the situation.

**Comment by jkotas** [comment]
> I think we also might be able to move to QCalls with SuppressGCTransition instead of FCalls without HELPER_METHOD_FRAME for the perf critical ones, but that's a bridge to examine and consider crossing once we've removed HELPER_METHOD_FRAME from the situation.

I agree. It would require modifying rules for how GC tracked pointers are passed to and from P/Invokes and we would need to deal with managed/unmanaged calling convention differences somehow. I think it would be just a cosmetic change at the end. It would not be a significant simplification of the system (compared to getting rid of HELPER_METHOD_FRAMEs).

**Comment by huoyaoyuan** [comment]
I believe this was closed unintentionally 😄 

In addition, I'd like to see some documentation about the more and more unmanaged-like unsafe code in CoreLib. I'm still learning and trying case by case.

**Comment by jkotas** [comment]
> I'd like to see some documentation about the more and more unmanaged-like unsafe code in CoreLib.

What would you like to see? The documentation about QCalls and related topics is in https://github.com/dotnet/runtime/blob/main/docs/design/coreclr/botr/corelib.md .

**Comment by huoyaoyuan** [comment]
The existing documentation is about "manual managed" code. Nowadays there are more "manual unmanaged" code in CoreLib, for example, dealing with MethodTable and manipulating GC refs as bytes.

**Comment by jkotas** [comment]
Feel free to propose documentation update in a PR.

The examples do not look unique to CoreLib. MethodTable is an example of an unmanaged resource with lifetime associated with managed object. These situations occur in interop quite frequently. GC.KeepAlive documentation has an example with COM.

**Comment by AaronRobinsonMSFT** [comment]
See for some interesting notes: https://github.com/dotnet/runtime/issues/110767

**Comment by AaronRobinsonMSFT** [comment]
From: https://github.com/dotnet/runtime/issues/110767#issuecomment-2547530773

> The way @VSadov consolidated lock primitives with NativeAOT, if we continue on that path, maybe it is possible to share the entire implementation with AOT (which has this HCall written in C#).

@am11 Agree. The goal for any converted HCalls, likely needed in assembly, is that they could be shared with native AOT.

> There is also a related JIT_Patchpoint HCall, which ends up using VirtualUnwindToFirstManagedCallFrame, not sure if there is a better structured way to tackle it without the libunwind dependency but would be nice to switch to one if it exists.

Yep, the `JIT_Patchpoint` calls `JIT_Patchpoint_Framed`. This is a common pattern when an FCall/HCall was frameless and then it needed one a new function was written with the `_Framed` or `_Helper` suffix.

**Comment by am11** [comment]
`IL_Throw`, `IL_Rethrow` and `JIT_Box` can be checked.

**Comment by am11** [comment]
@AaronRobinsonMSFT JIT_Box was https://github.com/dotnet/runtime/pull/115134

**Comment by filipnavara** [comment]
Just a heads up, I will submit a PR to convert the allocation helpers soon. I have about half of the platforms done.

**Comment by AaronRobinsonMSFT** [comment]
@filipnavara That is great news. Thanks!

**Comment by filipnavara** [comment]
We are one PR away from wrapping up this effort!
